packages/gatsby/src/schema/schema.js

```diff
@@ -0,0 +1,362 @@
+    nodeStore,
+    typeConflictReporter,
+    typeMapping,
+    parentSpan,
+  })
+  processTypeComposer({ schemaComposer, typeComposer, nodeStore, parentSpan })
+  return schemaComposer.buildSchema()
+}
+
+module.exports = {
+  buildSchema,
+  rebuildSchemaWithSitePage,
+}
+
+const updateSchemaComposer = async ({
+  schemaComposer,
+  nodeStore,
+  types,
+  typeMapping,
+  thirdPartySchemas,
+  typeConflictReporter,
+  parentSpan,
+}) => {
+  await addTypes({ schemaComposer, parentSpan, types })
+  await addInferredTypes({
+    schemaComposer,
+    nodeStore,
+    typeConflictReporter,
+    typeMapping,
+    parentSpan,
+  })
+  await addSetFieldsOnGraphQLNodeTypeFields({
+    schemaComposer,
+    nodeStore,
+    parentSpan,
+  })
+  await Promise.all(
+    Array.from(schemaComposer.values()).map(typeComposer =>
+      processTypeComposer({
+        schemaComposer,
+        typeComposer,
+        nodeStore,
+        parentSpan,
+      })
+    )
+  )
+  await addThirdPartySchemas({ schemaComposer, thirdPartySchemas, parentSpan })
+  await addCustomResolveFunctions({ schemaComposer, parentSpan })
+}
+
+const processTypeComposer = async ({
+  schemaComposer,
+  typeComposer,
+  nodeStore,
+  parentSpan,
+}) => {
+  if (
+    typeComposer instanceof schemaComposer.TypeComposer &&
+    typeComposer.hasInterface(`Node`)
+  ) {
+    await addNodeInterfaceFields({ schemaComposer, typeComposer, parentSpan })
+    await addResolvers({ schemaComposer, typeComposer, parentSpan })
+    await addConvenienceChildrenFields({
+      schemaComposer,
+      typeComposer,
+      nodeStore,
+      parentSpan,
+    })
+    await addTypeToRootQuery({ schemaComposer, typeComposer, parentSpan })
+  }
+}
+
+const addTypes = ({ schemaComposer, types, parentSpan }) => {
+  types.forEach(typeOrTypeDef => {
+    if (typeof typeOrTypeDef === `string`) {
+      const addedTypes = schemaComposer.addTypeDefs(typeOrTypeDef)
+      addedTypes.forEach(type => {
+        const { GraphQLInterfaceType, GraphQLUnionType } = require(`graphql`)
+        let typeComposer
+        if (type instanceof GraphQLInterfaceType) {
+          typeComposer = schemaComposer.getOrCreateIFTC(type.name)
+        } else if (type instanceof GraphQLUnionType) {
+          typeComposer = schemaComposer.getOrCreateUTC(type.name)
+        }
+        if (typeComposer) {
+          typeComposer.setResolveType(node => node.internal.type)
+          schemaComposer.addSchemaMustHaveType(typeComposer)
+        }
+      })
+    } else {
+      schemaComposer.add(typeOrTypeDef)
+    }
+  })
+}
+
+const addSetFieldsOnGraphQLNodeTypeFields = ({
+  schemaComposer,
+  nodeStore,
+  parentSpan,
+}) =>
+  Promise.all(
+    Array.from(schemaComposer.values()).map(async tc => {
+      if (
+        tc instanceof schemaComposer.TypeComposer &&
+        tc.hasInterface(`Node`)
+      ) {
+        const typeName = tc.getTypeName()
+        const [fields] = await apiRunner(`setFieldsOnGraphQLNodeType`, {
```

Comment:
My mistake - I think the array returned by `setFields...` can have more than one entry. I'l doublecheck and fix this.

Reply:
`apiRunner` will return array because multiple plugins can implement same API - I.e. both plugin A and plugin B extends `MarkdownRemark` type, but they will be separate items in returned array.